src/eval.ml

```diff
@@ -218,11 +258,92 @@ and eq_modulo : term -> term -> bool = fun a b ->
   if !log_enabled then log_eqmd (r_or_g res "%a == %a") pp a pp b; res
 
+(** [tree_walk t c s] tries to match stack [s] against tree [t] of capacity
+    [c]. *)
+and tree_walk : Dtree.t -> int -> term list -> term option =
+  fun tree capa stk ->
+  let vars = Array.make capa Kind in (* dummy terms *)
+  let module R = Dtree.ReductionStack in
+  let stk = R.of_list stk in
+  (* [walk t s c] where [s] is the stack of terms to match and [c] the cursor
+     indicating where to write in the [env] array described in {!module:Terms}
+     as the environment of the RHS during matching. *)
+  let rec walk tree stk cursor =
+    match tree with
+    | Fail                                   -> None
+    | Leaf(env_builder, act)                 ->
+        (* Allocate an environment for the action. *)
+        let env = Array.make (Bindlib.mbinder_arity act) TE_None in
+        (* Retrieve terms needed in the action from the [vars] array. *)
+        let fn pos slot =
+          let t = unfold vars.(pos) in
+          let b = Bindlib.raw_mbinder [||] [||] 0 mkfree (fun _ -> t) in
+          env.(slot) <- TE_Some(b)
+        in
+        IntMap.iter fn env_builder;
+        (* Actually perform the action. *)
+        Some(Bindlib.msubst act env)
+    | Node({swap; children; store; default}) ->
+        (* Quit if stack is too short. *)
+        if R.is_empty stk || swap >= R.length stk then None else
+        (* Pick the right term in the stack. *)
+        let (left, examined, right) = R.destruct stk swap in
+        (* Store hd of stack if needed *)
+        if store then begin
+          if !log_enabled then log_eval "(node) storing [%a]" pp examined ;
+          vars.(cursor) <- examined
+        end ;
+        let cursor = if store then succ cursor else cursor in
+        let r_ex = to_tref examined in
+        (* Fetch the right subtree and the new stack *)
+        (* [choose t] chooses a tree among {!val:children} when term [t] is
+           examined and returns the new head of stack. *)
+        let choose t : tree option * term list =
+          let h, args = get_args t in
+          match h with
+          | Symb(s, _)  ->
+            let args = List.map ensure_tref args in
+            let c_ari = List.length args in
+            r_ex := Some(add_args h args) ;
+            let cons = { c_sym = s.sym_name ; c_mod = s.sym_path ; c_ari} in
+            let matched = TcMap.find_opt cons children in
+            if matched = None then (default, []) else (matched, args)
+          | Abst(_, _) -> assert false
+          | Meta(_, _) -> assert false
+          | _          -> assert false
+        in
+        let (matched, args) = if TcMap.is_empty children
+          then (default, []) else choose (whnf examined) in
+        let stk = R.restruct left args right in
+        Option.bind (fun tr -> walk tr stk cursor) matched
+    | Fetch(store, next)                     ->
+        let left, examined, right = R.destruct stk 0 in
```

Comment:
I guess that, in this case, left is [], examined is the head, and right the tail, right? But it could raise Not_found if the stk is empty, and Not_found is not caught here...